Sources/AsyncKit/ConnectionPool/EventLoopConnectionPool.swift

```diff
@@ -20,85 +20,79 @@ import NIOCore
-    
+    ///
+    /// > Note: Any connection in this list may have become invalid since its last use.
+    private var available: Deque<PrunableConnection<Source.Connection>>
+
     /// Current active connection count.
-    public private(set) var activeConnections: Int
-    
-    /// All requests for a connection that were unable to be fulfilled
-    /// due to max connection limit having been reached.
-    private var waiters: CircularBuffer<(Logger, EventLoopPromise<Source.Connection>)>
-    
+    private var activeConnections: Int
+
+    /// Connection requests waiting to be fulfilled due to pool exhaustion.
+    private var waiters: OrderedDictionary<Int, WaitlistItem>
+
     /// If `true`, this storage has been shutdown.
     private var didShutdown: Bool
-    
+
     /// For lifecycle logs.
     public let logger: Logger
     
-    private let lock: Lock
-    
     /// Creates a new `EventLoopConnectionPool`.
     ///
     ///     let pool = EventLoopConnectionPool(...)
     ///     pool.withConnection(...) { conn in
     ///         // use conn
     ///     }
     ///
-    /// - parameters:
-    ///     - source: Creates new connections when needed.
-    ///     - maxConnections: Limits the number of connections that can be open.
-    ///                       Defaults to 1.
-    ///     - requestTimeout: Timeout for requesting a new connection.
-    ///                       Defaults to 10 seconds.
-    ///     - logger: For lifecycle logs.
-    ///     - on: Event loop.
+    /// - Parameters:
+    ///   - source: Creates new connections when needed.
+    ///   - maxConnections: Limits the number of connections that can be open. Defaults to 1.
+    ///   - requestTimeout: Timeout for requesting a new connection. Defaults to 10 seconds.
+    ///   - logger: For lifecycle logs.
+    ///   - on: Event loop.
     public init(
         source: Source,
         maxConnections: Int,
         requestTimeout: TimeAmount = .seconds(10),
+        pruneInterval: TimeAmount = .seconds(60),
+        maxIdleTimeBeforePruning: TimeAmount = .seconds(120),
         logger: Logger = .init(label: "codes.vapor.pool"),
-        on eventLoop: EventLoop,
-        pruneInterval: TimeInterval = 60,
-        maxIdleTimeBeforePrunning: TimeInterval = 120
+        on eventLoop: EventLoop
     ) {
         self.source = source
         self.maxConnections = maxConnections
         self.requestTimeout = requestTimeout
         self.logger = logger
         self.eventLoop = eventLoop
-        self.available = .init(initialCapacity: maxConnections)
+        self.available = .init(minimumCapacity: maxConnections)
         self.activeConnections = 0
-        self.waiters = .init()
+        self.waiters = .init(minimumCapacity: maxConnections << 2)
         self.didShutdown = false
-        self.pruneInterval = pruneInterval
-        self.maxIdleTimeBeforePrunning = maxIdleTimeBeforePrunning
-        self.lock = .init()
```

Comment:
Don't need these 3 stored in the class.
The lock isn't needed since we ensure everything is don eon the eventloop.